accepted/future-releases/extension-types/feature-specification.md

```diff
@@ -0,0 +1,1098 @@
+following the normal rules for constant constructors.
+
+A compile-time error occurs if an extension type constructor includes a
+superinitializer. *That is, a term of the form `super(...)` or
+`super.id(...)` as the last element of the initializer list.*
+
+A compile-time error occurs if an extension type constructor declares a
+super parameter. *For instance, `Name(super.x);`.*
+
+*In the body of a generative extension type constructor, the static type of
+`this` is the same as it is in any instance member of the extension type
+declaration, that is, `Name<X1 .. Xk>`, where `X1 .. Xk` are the type
+parameters declared by `Name`.*
+
+An instance creation expression of the form
+<code>Name&lt;T<sub>1</sub>, .. T<sub>s</sub>&gt;(...)</code>
+or
+<code>Name&lt;T<sub>1</sub>, .. T<sub>s</sub>&gt;.name(...)</code>
+is used to invoke these constructors, and the type of such an expression is
+<code>Name&lt;T<sub>1</sub>, .. T<sub>s</sub>&gt;</code>.
+
+*In short, extension type constructors appear to be very similar to
+constructors in classes, and they correspond to the situation
+where the enclosing class has a single, non-late, final instance
+variable, which is initialized according to the normal rules for
+constructors (in particular, it can occur by means of `this.id`, or in
+an initializer list, or by an initializing expression in the
+declaration itself, but it is an error if it does not occur at all).*
+
+An extension type `V` used as an expression (*a type literal*) is allowed
+and has static type `Type`.
+
+An extension type declaration _DV_ can have the type modifier
+`final`. In this case it is a compile-time error for another extension type
+declaration _DV2_ to have _DV_ as a superinterface.
+
+*As the grammar shows, any occurrence of the keywords `abstract`,
+`base`, `interface`, `sealed`, or `mixin` in an extension type
+declaration is a syntax error.*
+
+
+### Composing Extension Types
+
+This section describes the effect of including a clause derived from
+`<interfaces>` in an extension type declaration. We use the phrase
+_the implements clause_ to refer to this clause.
+
+*The rationale is that the set of members and member implementations of a
+given extension type may need to overlap with that of other extension type
+declarations. The implements clause allows for implementation reuse by
+putting some shared members in an extension type `V`, and including `V` in
+the implements clause of several extension type declarations
+<code>V<sub>1</sub> .. V<sub>k</sub></code>, thus "inheriting" the members
+of `V` into all of <code>V<sub>1</sub> .. V<sub>k</sub></code> without code
+duplication.*
+
+*The reason why this mechanism uses the keyword `implements` rather
+than `extends` to declare a relation that involves "inheritance" is
+that it has a similar semantics as that of extension members (in that
+they are statically resolved, and each member is applicable to every
+subtype).*
```

Comment:
The big difference is that a mixin application is ordered, and members with the same name will have the implementation that is declared in the most specific (latest applied) mixin. With extension method, you have unlimited symmetric combination ("multiple inheritance"), and then it's an error if there are name clashes on equally specific extensions.

With extension types, we insist on resolving name clashes explicitly by a redeclaration in the most specific extension type. We don't have a notion of `super` invocations, so the developer who just wants to select one specific implementation must write a forwarder (something like `void m() => (this as V3).m();`). We could add some notion of `super` later on, and this could allow those forwarders to be much more concise (avoiding the duplication of long parameter lists and such), but we don't have that kind of mechanism now. It should be no problem to start using it later, if we add it.

`implements` is relevant because extension methods are applicable to any receiver whose type is a subtype of the on-type:

```dart
extension type V2(int i) { void m() {}}
extension type V1(int i) implements V2 {}

extension E2 on V2 { void n() {}}
extension E1 on V1 {} // We don't need this one, but it doesn't hurt either.

void main() {
  var v1 = V1(42);
  v1.m(); // `V2.m` is applicable to `v1` because `V1 <: V2`.
  v1.n(); // `E2.n` is applicable to `v1` because `V1 <: V2`.
}
```

Reply:
Ok, that makes sense, and is a good reason not to use `with`. `implements` isn't great either, but I can't think of a better word at the moment, and I'm convinced that `implements` might be the least-bad option. (Although it might be worth playing around with other words, if you haven't already, such as `subtypes` or `enhances` to try to find something that doesn't already have a semantic different than what we're doing here.)

Reply:
We've played around with all these words for years. I'd like to see something landing firmly. But `implements` has been justified by this relationship with extension method resolution for many months now, so I think that explanation sticks.